### PR TITLE
Per-user KDF ratchets for message protection

### PR DESCRIPTION
**Pull request by Scratch-net**
This PR is a work towards message protection for MLS. It specifies separate KDF ratchets per every user to eliminate possible nonce reuse.

### CODE REVIEW

File: draft-barnes-mls-protocol.md [hunk 1]
```diff
@@ -1151,6 +1151,67 @@ failing to get their proposal accepted.
 
 [[OPEN ISSUE: Another possibility here is batching + deterministic selection.]]
 
+## Client Secrets
+
+To avoid conflicts when using the same Group secret, each Member MUST
+have his unique Client secret derived from the root key of the ratcheting tree
```
[the remark this answers is not in the document]
Comment (Scratch-net) [reply draft-barnes-mls-protocol.md]: Agree )

File: draft-barnes-mls-protocol.md [hunk 2]
```diff
@@ -1151,6 +1151,67 @@ failing to get their proposal accepted.
...
+and client's index in the ratcheting tree.
+Msg serves the domain separation purpose and should be a non-empty constant string.
+
+This is achieved by using HKDF:
+
+~~~~~
+Derive-Client-Secret(Secret, Index, Msg) =
```
[the remark this answers is not in the document]
Comment (Scratch-net) [reply draft-barnes-mls-protocol.md]: I've replaced Secret with Epoch Secret to more precise about what initially goes into HKDF

File: draft-barnes-mls-protocol.md [hunk 3]
```diff
@@ -1151,6 +1151,67 @@ failing to get their proposal accepted.
...
+    uint32 tree_index = Index;
+    opaque message<1..2^16-1> = Msg;
+} ClientHkdfLabel;
+~~~~~
+
+HKDF Output is used as the Client Secret
+
+## Generating message keys using KDF chains
+We adopt {{doubleratchet}} for MLS in a way that Diffie-Hellman ratchet is replaced
+by {{art}} and each Member has his own sending KDF ratchet.
```
[the remark this answers is not in the document]
Comment (Scratch-net) [reply draft-barnes-mls-protocol.md]: Fixed

**Comment by Scratch-net** [comment]
Hey @beurdouche, thanks! 
Made amends as per your comments, see my latest commit. The scheme still may need some love but overall I hope it became better )

File: draft-barnes-mls-protocol.md [hunk 4]
```diff
@@ -1151,6 +1151,58 @@ failing to get their proposal accepted.
...
+
+~~~~~
+Derive-Client-Secret(Secret, Index, Label) =
+     HKDF-Expand(Secret, ClientHkdfLabel, Length)
+
+Where ClientHkdfLabel is specified as:
+
+struct {
+    uint32 tree_index = Index;
+    opaque label<1..2255> = Label;
```
Comment (beurdouche) [inline draft-barnes-mls-protocol.md]: Apple Keyboard ? When I press 9, 99 appears... ;P

**Comment by beurdouche** [comment]
Looks Good, Thanks Alexey ! : )

File: draft-barnes-mls-protocol.md (hunk 4 above)
Comment (Scratch-net) [reply draft-barnes-mls-protocol.md]: Ah, mess, no, windows ) fill fix

**Comment by Scratch-net** [comment]
fixed & squashed

**Comment by bifurcation** [comment]
This was obsoleted by #54

**Pull request closed without merging**